### ISSUE TITLE
fe/mapping_manifold_06 fails with floating point exception

### ISSUE THREAD

**Issue by tamiko** [dealii/dealii#2763]
The test ```fe/mapping_manifold_06``` fails in 3D with a floating point exception. The issue is that
in ```mapping_manifold_06.cc``` the call
```
67           fe_v.reinit(cell, f);
```
fails at ```source/fe/mapping_manifold.cc```
```
780                   output_data.normal_vectors[i] = Point<spacedim>(output_data.boundary_forms[i] /
781                                                                   output_data.boundary_forms[i].norm());
```
because the ```norm()``` is 0. That's scary. Further, for 2D all ```output_data.boundary_forms[i].norm()``` report the same number, whereas in 3D we get very different results.

I don't know what's going on.

@luca-heltai @ESeNonFossiIo @asartori86

**Comment by tamiko** [comment]
ugly debug code:
```
diff --git a/source/fe/mapping_manifold.cc b/source/fe/mapping_manifold.cc
index 9bdb462..a92f37d 100644
--- a/source/fe/mapping_manifold.cc
+++ b/source/fe/mapping_manifold.cc
@@ -762,8 +762,14 @@ namespace internal
 
           if (update_flags & (update_normal_vectors
                               | update_JxW_values))
+          {
+            std::cout << "############################" << std::endl;
             for (unsigned int i=0; i<output_data.boundary_forms.size(); ++i)
               {
+                std::cout << output_data.boundary_forms[i].norm() << " <<< ---------------------------------" << std::endl;
+                Assert(output_data.boundary_forms[i].norm() > 1.e-16,
+                       ExcInternalError());
+
                 if (update_flags & update_JxW_values)
                   {
                     output_data.JxW_values[i] = output_data.boundary_forms[i].norm() * weights[i];
@@ -780,6 +786,7 @@ namespace internal
                   output_data.normal_vectors[i] = Point<spacedim>(output_data.boundary_forms[i] /
                                                                   output_data.boundary_forms[i].norm());
               }
+          }
 
           if (update_flags & update_jacobians)
             for (unsigned int point=0; point<n_q_points; ++point)
diff --git a/tests/fe/mapping_manifold_06.cc b/tests/fe/mapping_manifold_06.cc
index cdd80c1..9894a05 100644
--- a/tests/fe/mapping_manifold_06.cc
+++ b/tests/fe/mapping_manifold_06.cc
@@ -27,6 +27,8 @@
 #include <deal.II/fe/fe_values.h>
 #include <deal.II/grid/manifold_lib.h>
 
+#include <deal.II/grid/grid_out.h>
+
 template<int dim, int spacedim>
 void test()
 {
```

**Comment by bangerth** [comment]
Looking into it.

**Comment by bangerth** [comment]
Progress report: We fail to compute the Jacobian incorrectly, and the Jacobian is used to compute the tangential vectors from which we compute the boundary forms by a cross product.

Of course, this happens to happen for quadrature points located at the north and south poles of the sphere :-(

**Comment by bangerth** [comment]
OK, this took me a surprisingly long time to figure out (2.5 hours). Ultimately, the failure in this testcase is caused by the fact that we try to compute a tangent vector to a `SphericalManifold` object at one of the singular points of the transformation by which we fit the spherical manifold into the `ChartManifold` framework. This is illustrated by the following little test:
```
// ---------------------------------------------------------------------
//
// Copyright (C) 2016 by the deal.II authors
//
// This file is part of the deal.II library.
//
// The deal.II library is free software; you can use it, redistribute
// it, and/or modify it under the terms of the GNU Lesser General
// Public License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
// The full text of the license can be found in the file LICENSE at
// the top level of the deal.II distribution.
//
// ---------------------------------------------------------------------

// test tangent vectors to SphericalManifold at the poles

#include "../tests.h"

#include <deal.II/base/utilities.h>
#include <deal.II/grid/manifold_lib.h>


int
main()
{
  initlog();

  const SphericalManifold<3> manifold;

  // get tangent vectors at the south pole of the sphere in direction
  // of the meridional equator point and a point 90 degrees to the
  // east or west of that point. this should yield two tangent vectors
  // that are orthogonal to each other
  deallog << manifold.get_tangent_vector (Point<3>(0, 0, -1),
                                          Point<3>(1, 0, 0)) << std::endl
          << manifold.get_tangent_vector (Point<3>(0, 0, -1),
                                          Point<3>(0, 1, 0)) << std::endl;
}
```
What the program outputs, however, is this:
```
DEAL::1.57080 0.00000 1.92367e-16
DEAL::1.57080 1.92367e-16 1.92367e-16
```
In other words, it computes the *same vector* for both tangents. Consequently, nothing downstream will make sense. In particular, trying to compute the normal to the surface of the sphere at these points won't work because that would require taking the cross product of these two vectors (which is of course zero).

@luca-heltai -- I'm tempted to fix this whole issue by simply not implementing `SphericalManifold` based on `ChartManifold`. It just seems a bad idea to use a singular chart at the poles and surely we will find more issues such as this at one point or other if we tried to work around it somehow.

**Comment by tamiko** [comment]
@bangerth Thanks a lot for looking into this!

**Comment by luca-heltai** [comment]
Guys, this has been known for some months now, and a fix is in PR #2676, where SphericalManifold is no longer derived by ChartManifold, while the current SphericalManifold is moved to PolarManifold ( @ESeNonFossiIo )

**Comment by luca-heltai** [comment]
There are other issues in that PR, but if you wish, we can extract the relevant things, and get that one merged fast.

**Comment by bangerth** [comment]
Yes, if you could extract the relevant parts, that would be fine. 

#2676 is marked as "Work in Progress", so that's likely why nobody gives it a review. Do you want people to look at it?

**Comment by ESeNonFossiIo** [comment]
As @luca-heltai said there are still some issues to fix. 
Btw, I would appreciate any comments you can give to this PR. 
In this way the next PR will be simpler.

**Comment by bangerth** [comment]
OK. Please extract the parts the are relevant here for a separate pull request. Once merged, rebase onto master and I will comment on the rest then.

**Comment by bangerth** [comment]
@tamiko -- do you still have this problem after #2676 was merged? If not, can you please close?

**Comment by drwells** [comment]
I reran this on master and it looks like this has been fixed: if I use the same
modifications to `mapping_manifold.cc`:

```cpp
// starting at line 774
if (update_flags & (update_normal_vectors
                    | update_JxW_values))
  {
    std::cout << "#############################################" << std::endl;
    for (unsigned int i=0; i<output_data.boundary_forms.size(); ++i)
      {
        std::cout << output_data.boundary_forms[i].norm()
                  << " << ------------" << std::endl;
        Assert(output_data.boundary_forms[i].norm() > 1.0e-16,
               ExcInternalError());

        if (update_flags & update_JxW_values)
          {
            output_data.JxW_values[i] = output_data.boundary_forms[i].norm() * weights[i];

            if (subface_no!=numbers::invalid_unsigned_int)
              {
                const double area_ratio=GeometryInfo<dim>::subface_ratio(cell->subface_case(face_no),
                                                                         subface_no);
                output_data.JxW_values[i] *= area_ratio;
              }
          }

        if (update_flags & update_normal_vectors)
          output_data.normal_vectors[i] = Point<spacedim>(output_data.boundary_forms[i] /
                                                          output_data.boundary_forms[i].norm());
      }
  }
```

then the test's `stdout` log file shows that all of the norms are order one, so I believe that this has been fixed.

**Comment by bangerth** [comment]
Ah excellent. Care to make this into a testcase for the testsuite? (But let me know if you don't want to do that and I'll take care of it.)

**Comment by drwells** [comment]
I can do it (it should only take a minute)

**Comment by drwells** [comment]
To clarify: it looks like @luca-heltai added your test case on July 12 (a few days after you posted it). Since we already fixed this and have a test, I assume we can close this issue?

**Comment by drwells** [comment]
Since (AFAIK) this issue has been fixed and we have the test case in the test suite, I am going to go ahead and close this.
